src/unlock/admin_list.rs

```diff
@@ -0,0 +1,250 @@
+    let (_, root, compiled_proof) = if on {
+        build_smt_bl(hashes, hashes, &[])
+    } else {
+        build_smt_bl(hashes, &[], hashes)
+    };
+    (root, compiled_proof)
+}
+
+/// Build smt tree with old hashes, hashes on the tree and hashes not on the tree.
+/// # Panics
+///
+/// Panics if on hashes and off hashes have same element.
+pub fn build_smt_bl(
+    orig_hashes: &[[u8; 32]],
+    on_hashes: &[[u8; 32]],
+    off_hashes: &[[u8; 32]],
+) -> (Vec<[u8; 32]>, H256, Vec<u8>) {
+    assert!(
+        !on_hashes.iter().any(|x| off_hashes.contains(x)),
+        "hash can't not on and off at the same time, on:{:#?} off:{:#?}",
+        on_hashes,
+        off_hashes
+    );
+    let mut hashes = orig_hashes.to_vec();
+    hashes.extend(on_hashes);
+    hashes.sort_unstable();
+    hashes.dedup();
+    let hashes: Vec<[u8; 32]> = hashes
+        .into_iter()
+        .filter(|x| !off_hashes.contains(x))
+        .collect();
+    let pairs: Vec<(H256, H256)> = hashes.iter().map(map_exist).collect();
+    let proof_pairs: Vec<(H256, H256)> = on_hashes
+        .iter()
+        .map(map_noexist)
+        .chain(off_hashes.iter().map(map_noexist))
+        .collect();
+
+    let smt = new_smt(pairs);
+    let root = smt.root();
+
+    let proof = smt
+        .merkle_proof(proof_pairs.clone().into_iter().map(|(k, _)| k).collect())
+        .expect("gen proof.");
+    let compiled_proof = proof.compile(proof_pairs.clone()).expect("compile proof");
+    let test_on = compiled_proof
+        .verify::<CKBBlake2bHasher>(root, proof_pairs)
+        .expect("verify compiled proof.");
+
+    assert!(test_on == on_hashes.is_empty());
+    (hashes, *root, compiled_proof.into())
+}
+
+fn build_rc_rule(smt_root: &[u8; 32], is_black: bool, is_emergency: bool) -> Bytes {
+    let mut flags: u8 = 0;
+
+    if !is_black {
+        flags ^= WHITE_BLACK_LIST_MASK;
+    }
+    if is_emergency {
+        flags ^= EMERGENCY_HALT_MODE_MASK;
+    }
+    let rcrule = RCRuleBuilder::default()
+        .flags(flags.into())
+        .smt_root(smt_root.pack())
+        .build();
+    let res = RCDataBuilder::default()
+        .set(RCDataUnion::RCRule(rcrule))
+        .build();
+    res.as_bytes()
+}
+
+pub fn build_proofs(proofs: Vec<Vec<u8>>, proof_masks: Vec<u8>) -> SmtProofEntryVec {
+    assert_eq!(proofs.len(), proof_masks.len());
+
+    let mut builder = SmtProofEntryVecBuilder::default();
+    let iter = proofs.iter().zip(proof_masks.iter());
+    for (p, m) in iter {
+        let proof_builder = SmtProofBuilder::default()
+            .set(p.iter().map(|v| molecule::prelude::Byte::new(*v)).collect());
+
+        let temp = SmtProofEntryBuilder::default()
+            .proof(proof_builder.build())
+            .mask((*m).into());
+        builder = builder.push(temp.build());
+    }
+    builder.build()
+}
+
+pub fn generate_single_proof(on: bool, smt_key: &[[u8; 32]]) -> (Vec<u8>, Bytes) {
```

Comment:
Even though we just support white list for now, we also need to support black list in function arguments.